File: [0583][Medium][Delete_Operation_for_Two_Strings]/Delete_Operation_for_Two_Strings.py
# leetcode time     cost : 384 ms
# leetcode memory   cost : 17.3 MB 
# Time  Complexity: O(MN)
# Space Complexity: O(MN)
# solution 1, DP,
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n1 = len(word1)
        n2 = len(word2)
        if not n1:
            return n2
        if not n2:
            return n1            
        dp = [ [0]*(n2+1) for _ in range(n1+1)]
        for i in range(n1+1):
            dp[i][0] =  i
        for j in range(n2+1):
            dp[0][j] =  j
        for i in range(1,n1+1):
            for j in range(1,n2+1):
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = dp[i-1][j-1]
                else:
                    dp[i][j] = min(dp[i-1][j]+1,dp[i][j-1]+1)
        return dp[n1][n2]
    
def main():
    word1,word2 = "aeaswqsdaf","befdswerw"  # ans 13
    obj = Solution()
    res = obj.minDistance(word1,word2)
    print("return value is ",res)
    
if __name__ =='__main__':
    main()     

File: [0583][Medium][Delete_Operation_for_Two_Strings]/Delete_Operation_for_Two_Strings.java
// leetcode time     cost : max time exceed
// leetcode memory   cost : -- MB 
// Time  Complexity: O(2**max(m,n))
// Space Complexity: O(max(m,n))
// solution 1, LCS,
class Solution {
    public int minDistance(String s1, String s2) {
        return s1.length() + s2.length() - 2 * lcs(s1, s2, s1.length(), s2.length());
    }
    public int lcs(String s1, String s2, int m, int n) {
        if (m == 0 || n == 0)
            return 0;
        if (s1.charAt(m - 1) == s2.charAt(n - 1))
            return 1 + lcs(s1, s2, m - 1, n - 1);
        else
            return Math.max(lcs(s1, s2, m, n - 1), lcs(s1, s2, m - 1, n));
    }
}


public class Delete_Operation_for_Two_Strings {
    public static void main(String args[]) {
        String word1 = "aeaswqsdaf"; 
        String word2="befdswerw" ;// expect is 13,
        Solution Solution_obj = new Solution();
        int result = Solution_obj.minDistance(word1,word2);
        System.out.println("In java code,result is :" + result);
    }
}